src/graph.erl

```diff
@@ -0,0 +1,180 @@
+        false -> {error, vertex_not_existing};
+        {error, Reason} -> {error, Reason}
+    end.
+
+vertices_to_types(IdsMap, []) ->
+    {ok, IdsMap};
+vertices_to_types(IdsMap, [Key | Rest]) ->
+    {ok, Vertex} = get_vertex(Key),
+    case get_vertex(Key) of
+        {ok, Vertex} ->
+            Data = json_utils:decode(Vertex),
+            case maps:get(<<"type">>, Data) of
+                <<"user">> ->
+                    Users = maps:get(<<"users">>, IdsMap),
+                    vertices_to_types(maps:update(<<"users">>, Users ++ [maps:get(<<"id">>, Data)], IdsMap), Rest);
+                <<"group">> ->
+                    Groups = maps:get(<<"groups">>, IdsMap),
+                    vertices_to_types(maps:update(<<"groups">>, Groups ++ [maps:get(<<"id">>, Data)], IdsMap), Rest);
+                <<"space">> ->
+                    Spaces = maps:get(<<"spaces">>, IdsMap),
+                    vertices_to_types(maps:update(<<"spaces">>, Spaces ++ [maps:get(<<"id">>, Data)], IdsMap), Rest);
+                <<"provider">> ->
+                    Providers = maps:get(<<"providers">>, IdsMap),
+                    vertices_to_types(maps:update(<<"providers">>, Providers ++ [maps:get(<<"id">>, Data)], IdsMap), Rest)
+            end;
+        {error, Reason} -> {error, Reason}
+    end.
+
+-spec list_vertices() -> {ok, map()} | {error, any()}.
+list_vertices() ->
+    {ok, Keys} = persistence:keys("*"),
+    vertices_to_types(#{
+        <<"users">> => [],
+        <<"groups">> => [],
+        <<"spaces">> => [],
+        <<"providers">> => []
+    }, Keys).
+
+get_vertices_of_type(_Type, IdsList, []) ->
+    {ok, IdsList};
+get_vertices_of_type(Type, IdsList, [Key | Rest]) ->
+    case get_vertex(Key) of
+        {ok, Vertex} ->
+            Data = json_utils:decode(Vertex),
+            case maps:get(<<"type">>, Data) of
+                Type -> get_vertices_of_type(Type, IdsList ++ [maps:get(<<"id">>, Data)], Rest);
+                _ -> get_vertices_of_type(Type, IdsList, Rest)
+            end;
+        {error, Reason} -> {error, Reason}
+    end.
+
+-spec list_vertices(Type::binary()) -> {ok, list()} | {error, any()}.
+list_vertices(Type) ->
+    Keys = persistence:keys("*"),
+    get_vertices_of_type(Type, [], Keys).
+
+
+%% edges api - todo
+
+%%%% Parent, Child, Vertex to wszystko ID-ki, jesli wolisz mozesz zmienic nazwy na ParentId itd., jak uwazasz
+
+-spec create_edge(Parent::binary(), Child::binary(), Permissions::binary()) -> ok | {error, any()}.
+create_edge(_Arg0, _Arg1, _Arg2) ->
+    erlang:error(not_implemented).
+
+-spec update_edge(Parent::binary(), Child::binary(), Permissions::binary()) -> ok | {error, any()}.
+update_edge(_Arg0, _Arg1, _Arg2) ->
+    erlang:error(not_implemented).
+
+-spec remove_edge(Parent::binary(), Child::binary()) -> ok | {error, any()}.
+remove_edge(_Arg0, _Arg1) ->
+    erlang:error(not_implemented).
+
+-spec edge_exists(Parent::binary(), Child::binary()) -> boolean() | {error, any()}.
```

Comment:
To serio nie jest dobry pomysł. Jaki byłby do tego pattern match?

```erlang
case MaybeBoolean of
  A -> ...
  {error, Error} -> ...
end
```
Problem polega na tym, że `{error, any()}` pasuje do pierwszego patterna. Strasznie łatwo sobie zrobić buga przy takim typie. Dodatkowo bez case'a nie masz jak sprawdzić, czy operacja się powiodła.

Reply:
To Rerequest-review brzmi ciekawie, ogarnę to sobie, dzięki. Tymczasem co do boolean() / {ok, boolean()} -> ehh... może masz rację, muszę to jeszcze przemyśleć. Tzn. z jednej strony w takiej formie jak to wygląda teraz jest mi wygodniej zakładać, że jest zwracany czysty boolean, a jeśli jest zwracana krotka error to się to po prostu wywali i tyle, ale masz rację że to może być powodem jakich bugów na przyszłość. Hmm... niech ci będzie, zmieniam to